Replace deprecated GTest::GTest with GTest::gtest.

stellaraccident opened the pull request:
Fixes https://github.com/ROCm/TheRock/issues/419

stellaraccident committed:
Replace deprecated GTest::GTest with GTest::gtest.

Fixes https://github.com/ROCm/TheRock/issues/419

cgmb reviewed (commented):
We should switch because the old target is deprecated, but this requires a minimum version of CMake 3.20. If you don't mind, I'll tweak this to prefer `GTest::gtest` but fall back to `GTest::GTest`. I think there's still folks building solver on Ubuntu 22.04 using CMake 3.16. They'll eventually need to upgrade, but this is easy enough to accommodate.

cgmb committed:
Add fallback logic for CMake <3.20

GTest::gtest is only available in CMake 3.20 and newer. The rocSOLVER
library builds fine with GTest::GTest, so we can fall back to that
for older versions of CMake that lack the new target.

cgmb reviewed (approved):
This could also be done with the `$<TARGET_EXISTS>` [generator expression](https://cmake.org/cmake/help/v4.0/manual/cmake-generator-expressions.7.html#target-dependent-expressions) without adding any extra lines of code. I figured if/else might be easier to read and understand, but I'm not particularly attached to either method.

cgmb commented:
The generator expression alternative would replace `GTest::GTest` with `$<IF:$<TARGET_EXISTS:GTest::gtest>, GTest::gtest, GTest::GTest>`, making this a +2/-2 patch, like the original change. Anyway, I'm fine with whatever, so I'm setting this to auto-merge once it has sufficient reviews.

cgmb committed:
Use generator expressions for conciseness

stellaraccident commented:
> The generator expression alternative would replace `GTest::GTest` with `$<IF:$<TARGET_EXISTS:GTest::gtest>, GTest::gtest, GTest::GTest>`, making this a +2/-2 patch, like the original change. Anyway, I'm fine with whatever, so I'm setting this to auto-merge once it has sufficient reviews.

The generator is concise, which is good. If we have to do it (i.e. support <3.20), that seems like the least bad way to do it.

TorreZuk commented:
$<TARGET_EXISTS:tgt>[¶](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_EXISTS)
Added in version 3.12.

cgmb commented:
> The generator is concise, which is good. If we have to do it (i.e. support <3.20), that seems like the least bad way to do it.

Sure. I waffled on this and made the changes yesterday, but paused at the push stage only because using a generator adds subtleties that I don't entirely understand to `install(EXPORT)`. However, given that the targets using GTest aren't exported anyway, it doesn't matter.

I push the update later today.

> Added in version 3.12.

We're requiring at 3.13 anyway, so that should be fine.

cgmb commented:
Hmmm... the generation method seems to somehow accidentally be depending on behaviour newer than CMake 3.22. It works on CMake 3.25 and later (which is why it passed on my workstation and the math-ci). I'll take a look later.

TorreZuk commented:
Old form that is more robust probably better than using newer features just to save a couple lines.

cgmb committed:
Drop spaces from generator expressions

Support for spaces in the if generator expression appears to have been
added somewhere between CMake 3.22 and CMake 3.25. The IF generator
expressions without spaces should work on CMake 3.12 and newer. It was
tested as working on CMake 3.16 from Ubuntu 20.04.

cgmb commented:
It looks like removing the spaces fixed everything. The math-ci returned an UNSTABLE overall status, but every step of the build and test passed, so I'm presuming that's an infrastructure issue of some kind...

cgmb commented:
> The math-ci returned an UNSTABLE overall status, but every step of the build and test passed, so I'm presuming that's an infrastructure issue of some kind.

Our CI team confirmed this was some sort of issue communicating the build status to GitHub, and not an actual failure in the PR. I'm going to put this back for auto-merge..
